crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -475,6 +563,123 @@ impl WritebackCache {
+            debug!(
+                "object version unavailable for consumption: {:?} (current: {})",
+                obj_ref,
+                live_object.version()
+            );
+            return Err(SuiError::UserInputError {
+                error: UserInputError::ObjectVersionUnavailableForConsumption {
+                    provided_obj_ref: *obj_ref,
+                    current_version: live_object.version(),
+                },
+            });
+        }
+
+        let live_digest = live_object.digest();
+        if obj_ref.2 != live_digest {
+            debug!("object digest mismatch: {:?} vs {:?}", obj_ref, live_digest);
+            return Err(SuiError::UserInputError {
+                error: UserInputError::InvalidObjectDigest {
+                    object_id: obj_ref.0,
+                    expected_digest: live_digest,
+                },
+            });
+        }
+
+        Ok(())
+    }
+
+    fn clear_cached_locks(&self, locks: &[(ObjectRef, LockDetails)]) {
+        for (obj_ref, lock) in locks {
+            let entry = self.locked_transactions.entry(*obj_ref);
+            let occupied = match entry {
+                DashMapEntry::Vacant(_) => panic!("lock must exist"),
+                DashMapEntry::Occupied(occupied) => occupied,
+            };
+
+            if occupied.get() == lock {
+                trace!("clearing lock: {:?}", lock);
+                occupied.remove();
+            } else {
+                // this is impossible because the only case in which we overwrite a
+                // lock is when the lock is from a previous epoch. but we are holding
+                // execution_lock, so the epoch cannot have changed.
+                panic!("lock was changed since we set it");
+            }
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn acquire_transaction_locks(
+        &self,
+        epoch_store: &AuthorityPerEpochStore,
+        owned_input_objects: &[ObjectRef],
+        transaction: VerifiedSignedTransaction,
+    ) -> SuiResult {
+        let tx_digest = *transaction.digest();
+
+        let mut locks_to_write: Vec<(_, LockDetails)> =
+            Vec::with_capacity(owned_input_objects.len());
+
+        let object_ids = owned_input_objects.iter().map(|o| o.0).collect::<Vec<_>>();
+        let live_objects = self.multi_get_objects_must_exist(&object_ids)?;
+
+        // Note that this function does not have to operate atomically. If there are two racing threads,
+        // then they are either trying to lock the same transaction (in which case both will succeed),
+        // or they are trying to lock the same object in two different transactions, in which case
+        // the sender has equivocated, and we are under no obligation to help them form a cert.
+        for (obj_ref, live_object) in owned_input_objects.iter().zip(live_objects.into_iter()) {
+            match Self::verify_live_object(obj_ref, &live_object)
```

Comment:
It seems a waste to try set a bunch of locks and then realize a live object is invalid.
It may be better to first do a `.all()` to verify them all, and then set all locks.
We could still set locks one by one, no strong opinion there.

Reply:
good point, this simplified things a bit